src/key-derivation/hashing-password.md

```diff
@@ -0,0 +1,248 @@
+let expected =
+    "$argon2id$v=19$m=19456,t=2,p=1$YmFkIHNhbHQh$DqHGwv6NQV0VcaJi7jeF1E8IpfMXmXcpq4r2kKyqpXk";
+
+assert_eq!(expected, hash.to_string());
+
+// The verifier reads the salt and the parameters from the hash and verifies the result is equal
+Argon2::default().verify_password(password.as_bytes(), &hash).expect("invalid password");
+# }
+```
+
+Sometimes it may be necessary to be able to verify passwords that were hashed with different
+algorithms or parameters. This can be done using the traits in `password_hash`:
+
+```rust,noplaypen
+# fn main() {
+use password_hash::{PasswordHash, PasswordVerifier};
+
+use argon2::Argon2;
+use pbkdf2::Pbkdf2;
+use scrypt::Scrypt;
+
+// Can be: `$argon2`, `$pbkdf2`, or `$scrypt`
+let hash_string =
+    "$argon2id$v=19$m=19456,t=2,p=1$YmFkIHNhbHQh$DqHGwv6NQV0VcaJi7jeF1E8IpfMXmXcpq4r2kKyqpXk";
+let input_password = "password";
+
+let password_hash = PasswordHash::new(&hash_string).expect("invalid password hash");
+
+// Trait objects for algorithms to support
+let algs: &[&dyn PasswordVerifier] = &[&Argon2::default(), &Pbkdf2, &Scrypt];
+
+password_hash.verify_password(algs, input_password).expect("invalid password");
+# }
+```
+
+## Differences with Standard Hash Functions
+
+Compared with [standard hash functions](hashing-data/README.md), password hashes have a few features
+that make them well suited to this purpose:
+
+### Salt
+
+Imagine that we use SHA-256 to hash our password:
+
+> **WARNING**: DO NOT use SHA-256 as a password hashing algorithm. We use it in this example only
+> because it is easy to understand. See the [What should I use?] section at the top of this page for
+> better options.
+
+```rust,editable
+# fn main() {
+# use sha2::{Sha256, Digest};
+# use data_encoding::HEXLOWER;
+let input = "password";
+
+let mut hash = Sha256::new();
+hash.update(input.as_bytes());
+let result = hash.finalize();
+
+println!("Hash: {}", HEXLOWER.encode(&result));
+# }
+```
+
+This prints `5e884898da...`, which we would store in our database.
+
+Now some security event happens, and our database with all password hashes gets leaked. Now, the
+public has this information:
+
+```
++----------+------------+
+| username | pw_sha256  |
++----------+------------+
+| ferris   | 5e884898da | <- note this hash
+| curie    | 437dd76609 |
+| turing   | fc3a03a63b |
++----------+------------+
+```
+
+An attacker is ready, and has calculated a table of common password hashes:
+
+```
++----------+------------+
+| input    | sha_256    |
++----------+------------+
+| 1234     | 03ac674216 |
+| abcd     | 84e73dc50f |
+| password | 5e884898da | <- note this hash
++----------+------------+
+```
+
+Bingo! user `ferris` has password hash `5e884898da`, which means the password `password` will work
+to access ferris's account. The user is now compromised!
+
+This is a simple version of a rainbow attack. To prevent this, we do what is called "salting". This
```

Comment:
I suggest either define rainbow (history too) or better yet just delete the first sentence.  YMMV.

Reply:
I moved the rainbow thing to an appendix and simplified this section, take another look